### PR TITLE
Remove getCorrectedSize for fields

### LINKED ISSUE #2562: Remove getCorrectedSize
`getCorrectedSize` is a function defined on fields and icons to bridge the gap between old and new rendering: we cannot make changes to `getSize` to make new rendering work without breaking old rendering.

Many implementations of `getCorrectedSize` return the size of the field with some offset in X and Y.  We should be able to remove those offsets in pairs: in the code that made them necessary and in `getCorrectedSize`.

In the best case, we'll entirely remove `getCorrectedSize` and just use the bounding box of the `svgRoot`, as @BeksOmega prototyped here: https://github.com/google/blockly/compare/render/collab...BeksOmega:fixes/getCorrectedSize

This issue should not be worked on by external contributors until new rendering has landed.

### PR DESCRIPTION
**Pull request by rachel-fenichel**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

#2562

### Proposed Changes

Update field sizing to always describe all visible parts of the field, including the border rect if it exists.

### Reason for Changes

Make field sizing more consistent.
Note that this breaks the old renderer.

### Test Coverage
71/112 tests pass in LTR; 86/112 in RTL.  All field types pass, and the remaining test failures are due to other known bugs.

### Documentation

After I finish this + the followup work, will need to update the documentation on fields to include a description of what the size represents.

### Additional Information

Also related to #2530

Followup work: 
- Fix all fields that actually position themselves at -5 px, which may be all fields with a border rect, and remove `dealWithOffsetFields`
- Remove `getCorrectedSize` for icons
- Update documentation

### CODE REVIEW

File: core/field.js [hunk 1]
```diff
@@ -124,6 +129,20 @@ Blockly.Field.cacheReference_ = 0;
  */
 Blockly.Field.BORDER_RECT_DEFAULT_HEIGHT = 16;
 
+/**
+ * The default height of the text element on any field.
+ * @type {number}
+ * @package
+ */
+Blockly.Field.TEXT_DEFAULT_HEIGHT = 12.5;
+
```
Comment (samelhusseini) [inline core/field.js]: What is this number based on? Does it have any relationship with Blockly.FieldText.FONTSIZE?

Comment (samelhusseini) [reply core/field.js]: ie: if a developer changes the font size from 11pt to 12pt, would they have to fiddle around with this number?

File: core/field_colour.js [hunk 2]
```diff
@@ -78,15 +78,15 @@ Blockly.FieldColour.fromJson = function(options) {
...
+Blockly.FieldColour.DEFAULT_WIDTH = 16 + Blockly.Field.X_PADDING;
 
 /**
  * Default height of a colour field.
  * @type {number}
  * @private
  * @const
  */
-Blockly.FieldColour.DEFAULT_HEIGHT = 12;
+Blockly.FieldColour.DEFAULT_HEIGHT = Blockly.Field.BORDER_RECT_DEFAULT_HEIGHT;
```
Comment (RoboErikG) [inline core/field_colour.js]: Is losing the -1 intentional?

File: core/field.js (hunk 1 above)
Comment (rachel-fenichel) [reply core/field.js]: Yes, they would have to fiddle with it. As mentioned in one of the other bugs, we don't handle font size changes correctly.

Right now this just matches the previous value, though it often showed up as an implicit part of another sizing term.

File: core/field_image.js [hunk 3]
```diff
@@ -114,6 +116,14 @@ Blockly.FieldImage.prototype.EDITABLE = false;
  */
 Blockly.FieldImage.prototype.isDirty_ = false;
 
+/**
+ * Vertical padding below the image, which is included in the reported height of
+ * the field.
+ * @type {number}
+ * @private
+ */
+Blockly.FieldImage.Y_PADDING = 1;
```
Comment (RoboErikG) [inline core/field_image.js]: Should constants be at the top of the file?

File: core/field_image.js [hunk 4]
```diff
@@ -59,14 +59,16 @@ Blockly.FieldImage = function(src, width, height,
...
+  if (imageHeight <= 0 || imageWidth <= 0) {
     throw Error('Height and width values of an image field must be greater' +
       ' than 0.');
   }
-  this.size_ = new Blockly.utils.Size(this.width_,
-      this.height_ + 2 * Blockly.BlockSvg.INLINE_PADDING_Y);
+  // Store the image height, since it is different from the field height.
+  this.imageHeight_ = imageHeight;
+  this.size_ = new Blockly.utils.Size(imageWidth,
+      imageHeight + Blockly.FieldImage.Y_PADDING);
```
Comment (RoboErikG) [inline core/field_image.js]: Does this mean there's only 1/2 a pixel above and below an image or should this be 2*Y_PADDING?

File: core/field_colour.js (hunk 2 above)
Comment (rachel-fenichel) [reply core/field_colour.js]: Yes.  The field was reporting its height to be smaller than its actual bounds, in order to make the spacer below it smaller.  I don't want to do that--the height should be accurate--so I'm allowing it to have a slightly taller spacer below.

File: core/field_image.js (hunk 4 above)
Comment (rachel-fenichel) [reply core/field_image.js]: Only below the image.

File: core/field_image.js (hunk 3 above)
Comment (rachel-fenichel) [reply core/field_image.js]: We seem to have standardized on 
- constructor
- fromJson
- properties on the class
- properties on the prototype

I'm moving this above the properties on the prototype to match.  I think the goal was to keep fromJson as close to the constructor it calls as possible, but I'm okay with changing that for all fields at the same time.

Comment (RoboErikG) [reply core/field_image.js]: Yeah, that order seems right. This just needs to be above the prototype declarations.

Comment (rachel-fenichel) [reply core/field_image.js]: Done.

**Comment by rachel-fenichel** [comment]
I made the change Erik requested, and also removed dealWithOffsetFields, because it turned out to be easy.  PTAL the newest commit.

File: core/field_colour.js [hunk 5]
```diff
@@ -78,15 +78,15 @@ Blockly.FieldColour.fromJson = function(options) {
  * @private
  * @const
  */
-Blockly.FieldColour.DEFAULT_WIDTH = 16;
+Blockly.FieldColour.DEFAULT_WIDTH = 16 + Blockly.Field.X_PADDING;
```
Comment (BeksOmega) [inline core/field_colour.js]: Since there's no padding on a colour field, I think it would make more sense to set `DEFAULT_WIDTH = 26`.

**Comment by BeksOmega** [comment]
It seems confusing to me that elements get sized based on size_ and the size_ is set based on the size of elements. Is there a plan to reconcile this? Or is that further down the road?

File: core/field_colour.js (hunk 5 above)
Comment (rachel-fenichel) [reply core/field_colour.js]: Reasonable.  Done.

**Comment by rachel-fenichel** [comment]
@BeksOmega For the most part sizing is
- get the desired size of the element that is varying (e.g. text element or image element), often based on calling `getCachedWidth` 
- set the size, x and y attributes, etc. of the actual element
- update the border box size if needed
- if border box size is different from reported size, store border box size internally
- calculate reported size based on the values computed above and store in `this.size_`

So the values in `this.size_` should be based on the sizes of the elements, but not vice versa.

The exception is in `initView`, when I set the `size_` and then use it immediately, but that only happens once, not on every change

**Comment by BeksOmega** [comment]
> @BeksOmega For the most part sizing is...

Rad! If it's that easy to explain then I think it's great =)

**Pull request merged**